sstable: error swallowed in two-level NextPrefix

If the single-level iterator returns no KV due to an I/O error, the two-level iterator ignores the error and continues. Uncovered by #3051.

```
    random_test.go:111: error swallowed during NextPrefix("nua") with stack goroutine 22 [running]:
        runtime/debug.Stack()
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x64
        github.com/cockroachdb/pebble/sstable.TestIterator_RandomErrors.func1({0x14000198000?, {0x0?, 0x14000198000?}, 0x14000034000?})
        	/Users/jackson/go/src/github.com/cockroachdb/pebble/sstable/random_test.go:58 +0x50
        github.com/cockroachdb/pebble/vfs/errorfs.InjectorFunc.MaybeError(0x1400014f7d8?, {0x102c7c4a4?, {0x0?, 0x102952bec?}, 0x1400014f7e8?})
        	/Users/jackson/go/src/github.com/cockroachdb/pebble/vfs/errorfs/errorfs.go:152 +0x38
        github.com/cockroachdb/pebble/vfs/errorfs.(*errorFile).ReadAt(0x14000288150, {0x600002c28020, 0x79, 0x79}, 0x1400014f828?)
        	/Users/jackson/go/src/github.com/cockroachdb/pebble/vfs/errorfs/errorfs.go:452 +0x54
        github.com/cockroachdb/pebble/sstable.(*simpleReadable).ReadAt(0x1400014f848?, {0x10298a14c?, 0x2d08215834b62?}, {0x600002c28020?, 0x932fee80014f858?, 0x654d4b49?}, 0x1400014f858?)
        	/Users/jackson/go/src/github.com/cockroachdb/pebble/sstable/reader.go:1215 +0x3c
        github.com/cockroachdb/pebble/objstorage.(*NoopReadHandle).ReadAt(0x1400014f8b8?, {0x1030c8c10?, 0x103518940?}, {0x600002c28020?, 0x1?, 0x140000df120?}, 0x1400019f400?)
        	/Users/jackson/go/src/github.com/cockroachdb/pebble/objstorage/noop_readahead.go:24 +0x34
        github.com/cockroachdb/pebble/sstable.(*Reader).readBlock(0x140000ce000, {0x1030c8c10, 0x103518940}, {0x102da34c4?, 0x14000100b98?}, 0x0, {0x1030c8cf0, 0x14000288198}, 0x1400015a460, 0x0)
        	/Users/jackson/go/src/github.com/cockroachdb/pebble/sstable/reader.go:557 +0x1c8
        github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).loadBlock(0x14000100b00, 0x38?)
        	/Users/jackson/go/src/github.com/cockroachdb/pebble/sstable/reader_iter_single_lvl.go:402 +0x214
        github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).NextPrefix(0x14000100b00, {0x140000cdf38, 0x3, 0x17})
        	/Users/jackson/go/src/github.com/cockroachdb/pebble/sstable/reader_iter_single_lvl.go:1187 +0x1ec
        github.com/cockroachdb/pebble/sstable.(*twoLevelIterator).NextPrefix(0x14000100b00, {0x140000cdf38, 0x3, 0x17})
        	/Users/jackson/go/src/github.com/cockroachdb/pebble/sstable/reader_iter_two_lvl.go:810 +0x50
        github.com/cockroachdb/pebble/sstable.(*opGenerator).generateNextPrefix.func1({0x1030ccf70?, 0x14000100b00?})
        	/Users/jackson/go/src/github.com/cockroachdb/pebble/sstable/random_test.go:171 +0x3c
        github.com/cockroachdb/pebble/sstable.TestIterator_RandomErrors(0x14000198820)
        	/Users/jackson/go/src/github.com/cockroachdb/pebble/sstable/random_test.go:105 +0xcf0
        testing.tRunner(0x14000198820, 0x1030be828)
        	/usr/local/go/src/testing/testing.go:1595 +0xe8
        created by testing.(*T).Run in goroutine 1
        	/usr/local/go/src/testing/testing.go:1648 +0x33c
```

--- a/go.mod
+++ b/go.mod
@@ -6,6 +6,7 @@ require (
 	github.com/cespare/xxhash/v2 v2.2.0
 	github.com/cockroachdb/datadriven v1.0.3-0.20230413201302-be42291fc80f
 	github.com/cockroachdb/errors v1.11.1
+	github.com/cockroachdb/metamorphic v0.0.0-20231108215700-4ba948b56895
 	github.com/cockroachdb/redact v1.1.5
 	github.com/cockroachdb/tokenbucket v0.0.0-20230807174530-cc333fc44b06
 	github.com/ghemawat/stream v0.0.0-20171120220530-696b145b53b9
@@ -18,7 +19,7 @@ require (
 	github.com/prometheus/client_golang v1.12.0
 	github.com/prometheus/client_model v0.2.1-0.20210607210712-147c58e9608a
 	github.com/spf13/cobra v1.0.0
-	github.com/stretchr/testify v1.8.1
+	github.com/stretchr/testify v1.8.4
 	golang.org/x/exp v0.0.0-20230626212559-97b1e661b5df
 	golang.org/x/perf v0.0.0-20230113213139-801c7ef9e5c5
 	golang.org/x/sync v0.0.0-20220722155255-886fb9371eb4

--- a/sstable/random_test.go
+++ b/sstable/random_test.go
@@ -1,0 +1,395 @@
+// Copyright 2023 The LevelDB-Go and Pebble Authors. All rights reserved. Use
+// of this source code is governed by a BSD-style license that can be found in
+// the LICENSE file.
+
+package sstable
+
+import (
+	"fmt"
+	"math/rand"
+	"runtime/debug"
+	"slices"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/cockroachdb/metamorphic"
+	"github.com/cockroachdb/pebble/bloom"
+	"github.com/cockroachdb/pebble/internal/base"
+	"github.com/cockroachdb/pebble/internal/bytealloc"
+	"github.com/cockroachdb/pebble/internal/testkeys"
+	"github.com/cockroachdb/pebble/objstorage/objstorageprovider"
+	"github.com/cockroachdb/pebble/vfs"
+	"github.com/cockroachdb/pebble/vfs/errorfs"
+	"github.com/stretchr/testify/require"
+)
+
+// IterIterator_RandomErrors builds random sstables and runs random iterator
+// operations against them while randomly injecting errors. It ensures that if
+// an error is injected during an operation, operation surfaces the error to the
+// caller.
+func TestIterator_RandomErrors(t *testing.T) {
+	root := time.Now().UnixNano()
+	// Run the test a few times with various seeds for more consistent code
+	// coverage.
+	for i := int64(0); i < 50; i++ {
+		seed := root + i
+		t.Run(fmt.Sprintf("seed=%d", seed), func(t *testing.T) {
+			runErrorInjectionTest(t, seed)
+		})
+	}
+}
+
+func runErrorInjectionTest(t *testing.T, seed int64) {
+	t.Logf("seed %d", seed)
+	fs := vfs.NewMem()
+	f, err := fs.Create("random.sst")
+	require.NoError(t, err)
+	rng := rand.New(rand.NewSource(seed))
+	cfg := randomTableConfig{
+		wopts:     nil, /* leave to randomize */
+		keys:      testkeys.Alpha(3 + rng.Intn(2)),
+		keyCount:  10_000,
+		maxValLen: rng.Intn(64) + 1,
+		maxSuffix: rng.Int63n(95) + 5,
+		maxSeqNum: rng.Int63n(1000) + 10,
+		rng:       rng,
+	}
+	cfg.randomize()
+	_, err = buildRandomSSTable(f, cfg)
+	require.NoError(t, err)
+
+	f, err = fs.Open("random.sst")
+	require.NoError(t, err)
+	// Randomly inject errors into 25% of file operations. We use an
+	// errorfs.Toggle to avoid injecting errors until the file has been opened.
+	toggle := &errorfs.Toggle{Injector: errorfs.ErrInjected.If(errorfs.Randomly(0.25, seed))}
+	counter := &errorfs.Counter{Injector: toggle}
+	var stack []byte
+	f = errorfs.WrapFile(f, errorfs.InjectorFunc(func(op errorfs.Op) error {
+		err := counter.MaybeError(op)
+		if err != nil {
+			// Save the stack trace of the most recently injected error.
+			stack = debug.Stack()
+		}
+		return err
+	}))
+	readable, err := NewSimpleReadable(f)
+	require.NoError(t, err)
+	r, err := NewReader(readable, cfg.readerOpts())
+	require.NoError(t, err)
+	defer r.Close()
+
+	var filterer *BlockPropertiesFilterer
+	if rng.Float64() < 0.75 {
+		low, high := uint64(cfg.randSuffix()), uint64(cfg.randSuffix())
+		if low > high {
+			low, high = high, low
+		}
+		filterer = newBlockPropertiesFilterer([]BlockPropertyFilter{
+			NewTestKeysBlockPropertyFilter(low, high),
+		}, nil)
+	}
+
+	// TOOD(jackson): NewIterWithBlockPropertyFilters returns an iterator over
+	// point keys only. Should we add variants of this test that run random
+	// operations on the range deletion and range key iterators?
+	var stats base.InternalIteratorStats
+	it, err := r.NewIterWithBlockPropertyFilters(
+		nil /* lower TODO */, nil, /* upper TODO */
+		filterer,
+		rng.Intn(2) == 1, /* use filter block */
+		&stats,
+		CategoryAndQoS{},
+		nil, /* CategoryStatsCollector */
+		TrivialReaderProvider{r},
+	)
+	require.NoError(t, err)
+	defer it.Close()
+
+	// Begin injecting errors.
+	toggle.On()
+
+	ops := opRunner{randomTableConfig: cfg, it: it}
+	nextOp := metamorphic.Weighted[func() bool]{
+		{Item: ops.runSeekGE, Weight: 2},
+		{Item: ops.runSeekPrefixGE, Weight: 2},
+		{Item: ops.runSeekLT, Weight: 2},
+		{Item: ops.runFirst, Weight: 1},
+		{Item: ops.runLast, Weight: 1},
+		{Item: ops.runNext, Weight: 5},
+		{Item: ops.runNextPrefix, Weight: 5},
+		{Item: ops.runPrev, Weight: 5},
+	}.RandomDeck(rng)
+
+	for i := 0; i < 1000; i++ {
+		beforeCount := counter.Load()
+
+		// nextOp returns a function that *may* run the operation. If the
+		// current test state makes the operation an invalid operation, the the
+		// function returns `false` indicating it was not run. If the operation
+		// is a valid operation and was performed, `opFunc` returns true.
+		//
+		// This loop will run exactly 1 operation, skipping randomly chosen
+		// operations that cannot be run on an iterator in its current state.
+		for opFunc := nextOp(); !opFunc(); {
+			opFunc = nextOp()
+		}
+
+		t.Logf("%s = %s [err = %v]", ops.latestOpDesc, ops.k, it.Error())
+		afterCount := counter.Load()
+		// TODO(jackson): Consider running all commands against a parallel
+		// iterator constructed over a sstable containing the same data in a
+		// standard construction (eg, typical block sizes) and no error
+		// injection. Then we can assert the results are identical.
+
+		if afterCount > beforeCount {
+			if ops.k != nil || it.Error() == nil {
+				t.Errorf("error swallowed during %s with stack %s",
+					ops.latestOpDesc, string(stack))
+			}
+		}
+	}
+}
+
+type opRunner struct {
+	randomTableConfig
+	it Iterator
+
+	latestOpDesc  string
+	latestSeekKey []byte
+	dir           int8
+	k             *base.InternalKey
+	v             base.LazyValue
+}
+
+func (r *opRunner) runSeekGE() bool {
+	k := r.randKey()
+	flags := base.SeekGEFlagsNone
+	if strings.HasPrefix(r.latestOpDesc, "SeekGE") &&
+		r.wopts.Comparer.Compare(k, r.latestSeekKey) > 0 && r.rng.Intn(2) == 1 {
+		flags = flags.EnableTrySeekUsingNext()
+	}
+	r.latestOpDesc = fmt.Sprintf("SeekGE(%q, TrySeekUsingNext()=%t)",
+		k, flags.TrySeekUsingNext())
+	r.latestSeekKey = k
+	r.k, r.v = r.it.SeekGE(k, base.SeekGEFlagsNone)
+	r.dir = +1
+	return true
+}
+
+func (r *opRunner) runSeekPrefixGE() bool {
+	k := r.randKey()
+	i := r.wopts.Comparer.Split(k)
+	flags := base.SeekGEFlagsNone
+	if strings.HasPrefix(r.latestOpDesc, "SeekPrefixGE") &&
+		r.wopts.Comparer.Compare(k, r.latestSeekKey) > 0 && r.rng.Intn(2) == 1 {
+		flags = flags.EnableTrySeekUsingNext()
+	}
+	r.latestOpDesc = fmt.Sprintf("SeekPrefixGE(%q, %q, TrySeekUsingNext()=%t)",
+		k[:i], k, flags.TrySeekUsingNext())
+	r.latestSeekKey = k
+	r.k, r.v = r.it.SeekPrefixGE(k[:i], k, flags)
+	r.dir = +1
+	return true
+}
+
+func (r *opRunner) runSeekLT() bool {
+	k := r.randKey()
+	r.latestOpDesc = fmt.Sprintf("SeekLT(%q)", k)
+	r.k, r.v = r.it.SeekLT(k, base.SeekLTFlagsNone)
+	r.dir = -1
+	return true
+}
+
+func (r *opRunner) runFirst() bool {
+	r.latestOpDesc = "First()"
+	r.k, r.v = r.it.First()
+	r.dir = +1
+	return true
+}
+
+func (r *opRunner) runLast() bool {
+	r.latestOpDesc = "Last()"
+	r.k, r.v = r.it.Last()
+	r.dir = -1
+	return true
+}
+
+func (r *opRunner) runNext() bool {
+	if r.dir == +1 && r.k == nil {
+		return false
+	}
+	r.latestOpDesc = "Next()"
+	r.k, r.v = r.it.Next()
+	r.dir = +1
+	return true
+}
+
+func (r *opRunner) runNextPrefix() bool {
+	// NextPrefix cannot be called to change directions or when an iterator is
+	// exhausted.
+	if r.dir == -1 || r.k == nil {
+		return false
+	}
+	p := r.k.UserKey[:r.wopts.Comparer.Split(r.k.UserKey)]
+	succKey := r.wopts.Comparer.ImmediateSuccessor(nil, p)
+	r.latestOpDesc = fmt.Sprintf("NextPrefix(%q)", succKey)
+	r.k, r.v = r.it.NextPrefix(succKey)
+	r.dir = +1
+	return true
+}
+
+func (r *opRunner) runPrev() bool {
+	if r.dir == -1 && r.k == nil {
+		return false
+	}
+	r.latestOpDesc = "Prev()"
+	r.k, r.v = r.it.Prev()
+	r.dir = -1
+	return true
+}
+
+type randomTableConfig struct {
+	wopts     *WriterOptions
+	keys      testkeys.Keyspace
+	keyCount  int
+	maxValLen int
+	maxSuffix int64
+	maxSeqNum int64
+	rng       *rand.Rand
+}
+
+func (cfg *randomTableConfig) readerOpts() ReaderOptions {
+	rOpts := ReaderOptions{
+		Comparer: testkeys.Comparer,
+		Filters:  map[string]FilterPolicy{},
+	}
+	if cfg.wopts.FilterPolicy != nil {
+		rOpts.Filters[cfg.wopts.FilterPolicy.Name()] = cfg.wopts.FilterPolicy
+	}
+	return rOpts
+}
+
+func (cfg *randomTableConfig) randomize() {
+	if cfg.wopts == nil {
+		cfg.wopts = &WriterOptions{
+			// Test all table formats in [TableFormatLevelDB, TableFormatMax].
+			TableFormat:             TableFormat(cfg.rng.Intn(int(TableFormatMax)) + 1),
+			BlockRestartInterval:    (1 << cfg.rng.Intn(6)),             // {1, 2, 4, ..., 32}
+			BlockSizeThreshold:      min(int(100*cfg.rng.Float64()), 1), // 1-100%
+			BlockSize:               (1 << cfg.rng.Intn(18)),            // {1, 2, 4, ..., 128 KiB}
+			IndexBlockSize:          (1 << cfg.rng.Intn(20)),            // {1, 2, 4, ..., 512 KiB}
+			BlockPropertyCollectors: nil,
+			WritingToLowestLevel:    cfg.rng.Intn(2) == 1,
+			Parallelism:             cfg.rng.Intn(2) == 1,
+		}
+		if v := cfg.rng.Intn(11); v > 0 {
+			cfg.wopts.FilterPolicy = bloom.FilterPolicy(v)
+		}
+		if cfg.wopts.TableFormat >= TableFormatPebblev1 && cfg.rng.Float64() < 0.75 {
+			cfg.wopts.BlockPropertyCollectors = append(cfg.wopts.BlockPropertyCollectors, NewTestKeysBlockPropertyCollector)
+		}
+	}
+	cfg.wopts.ensureDefaults()
+	cfg.wopts.Comparer = testkeys.Comparer
+}
+
+func (cfg *randomTableConfig) randKey() []byte {
+	return testkeys.KeyAt(cfg.keys, cfg.randKeyIdx(), cfg.randSuffix())
+}
+func (cfg *randomTableConfig) randSuffix() int64 { return cfg.rng.Int63n(cfg.maxSuffix + 1) }
+func (cfg *randomTableConfig) randKeyIdx() int64 { return cfg.rng.Int63n(cfg.keys.Count()) }
+
+func buildRandomSSTable(f vfs.File, cfg randomTableConfig) (*WriterMetadata, error) {
+	// Construct a weighted distribution of key kinds.
+	kinds := metamorphic.Weighted[base.InternalKeyKind]{
+		{Item: base.InternalKeyKindSet, Weight: 25},
+		{Item: base.InternalKeyKindSetWithDelete, Weight: 25},
+		{Item: base.InternalKeyKindDelete, Weight: 5},
+		{Item: base.InternalKeyKindSingleDelete, Weight: 2},
+		{Item: base.InternalKeyKindMerge, Weight: 1},
+	}
+	// TODO(jackson): Support writing range deletions and range keys.
+	// TestIterator_RandomErrors only reads through the point iterator, so those
+	// keys won't be visible regardless, but their existence should be benign.
+
+	// DELSIZED require Pebblev4 or later.
+	if cfg.wopts.TableFormat >= TableFormatPebblev4 {
+		kinds = append(kinds, metamorphic.ItemWeight[base.InternalKeyKind]{
+			Item: base.InternalKeyKindDeleteSized, Weight: 5,
+		})
+	}
+	nextRandomKind := kinds.RandomDeck(cfg.rng)
+
+	type keyID struct {
+		idx    int64
+		suffix int64
+		seqNum int64
+	}
+	keyMap := make(map[keyID]bool)
+	// Constrain the space we generate keys to the middle 90% of the keyspace.
+	// This helps exercise code paths that are only run when a seek key is
+	// beyond or before all index block entries.
+	sstKeys := cfg.keys.Slice(cfg.keys.Count()/20, cfg.keys.Count()-cfg.keys.Count()/20)
+	randomKey := func() keyID {
+		k := keyID{
+			idx:    cfg.rng.Int63n(sstKeys.Count()),
+			suffix: cfg.rng.Int63n(cfg.maxSuffix + 1),
+			seqNum: cfg.rng.Int63n(cfg.maxSeqNum + 1),
+		}
+		// If we've already generated this exact key, try again.
+		for keyMap[k] {
+			k = keyID{
+				idx:    cfg.rng.Int63n(sstKeys.Count()),
+				suffix: cfg.rng.Int63n(cfg.maxSuffix + 1),
+				seqNum: cfg.rng.Int63n(cfg.maxSeqNum + 1),
+			}
+		}
+		keyMap[k] = true
+		return k
+	}
+
+	var alloc bytealloc.A
+	keys := make([]base.InternalKey, cfg.keyCount)
+	for i := range keys {
+		keyID := randomKey()
+		kind := nextRandomKind()
+
+		var keyBuf []byte
+		alloc, keyBuf = alloc.Alloc(testkeys.SuffixLen(keyID.suffix) + cfg.keys.MaxLen())
+		n := testkeys.WriteKeyAt(keyBuf, sstKeys, keyID.idx, keyID.suffix)
+		keys[i] = base.MakeInternalKey(keyBuf[:n], uint64(keyID.seqNum), kind)
+	}
+	// The Writer requires the keys to be written in sorted order. Sort them.
+	slices.SortFunc(keys, func(a, b base.InternalKey) int {
+		return base.InternalCompare(testkeys.Comparer.Compare, a, b)
+	})
+
+	// Release keyMap and alloc; we don't need them and this function can be
+	// memory intensive.
+	keyMap = nil
+	alloc = nil
+
+	valueBuf := make([]byte, cfg.maxValLen)
+	w := NewWriter(objstorageprovider.NewFileWritable(f), *cfg.wopts)
+	for i := 0; i < len(keys); i++ {
+		var value []byte
+		switch keys[i].Kind() {
+		case base.InternalKeyKindSet, base.InternalKeyKindMerge:
+			value = valueBuf[:cfg.rng.Intn(cfg.maxValLen+1)]
+			cfg.rng.Read(value)
+		}
+		if err := w.Add(keys[i], value); err != nil {
+			return nil, err
+		}
+	}
+	if err := w.Close(); err != nil {
+		return nil, err
+	}
+	metadata, err := w.Metadata()
+	if err != nil {
+		return nil, err
+	}
+	return metadata, nil
+}

--- a/sstable/reader_iter_single_lvl.go
+++ b/sstable/reader_iter_single_lvl.go
@@ -1139,6 +1139,8 @@ func (i *singleLevelIterator) Next() (*InternalKey, base.LazyValue) {
 	i.boundsCmp = 0
 
 	if i.err != nil {
+		// TODO(jackson): Can this case be turned into a panic? Once an error is
+		// encountered, the iterator must be re-seeked.
 		return nil, base.LazyValue{}
 	}
 	if key, val := i.data.Next(); key != nil {
@@ -1164,6 +1166,8 @@ func (i *singleLevelIterator) NextPrefix(succKey []byte) (*InternalKey, base.Laz
 	// Seek optimization only applies until iterator is first positioned after SetBounds.
 	i.boundsCmp = 0
 	if i.err != nil {
+		// TODO(jackson): Can this case be turned into a panic? Once an error is
+		// encountered, the iterator must be re-seeked.
 		return nil, base.LazyValue{}
 	}
 	if key, val := i.data.NextPrefix(succKey); key != nil {

--- a/sstable/reader_iter_two_lvl.go
+++ b/sstable/reader_iter_two_lvl.go
@@ -791,6 +791,8 @@ func (i *twoLevelIterator) Next() (*InternalKey, base.LazyValue) {
 	i.boundsCmp = 0
 	i.maybeFilteredKeysTwoLevel = false
 	if i.err != nil {
+		// TODO(jackson): Can this case be turned into a panic? Once an error is
+		// encountered, the iterator must be re-seeked.
 		return nil, base.LazyValue{}
 	}
 	if key, val := i.singleLevelIterator.Next(); key != nil {
@@ -808,6 +810,8 @@ func (i *twoLevelIterator) NextPrefix(succKey []byte) (*InternalKey, base.LazyVa
 	i.boundsCmp = 0
 	i.maybeFilteredKeysTwoLevel = false
 	if i.err != nil {
+		// TODO(jackson): Can this case be turned into a panic? Once an error is
+		// encountered, the iterator must be re-seeked.
 		return nil, base.LazyValue{}
 	}
 	if key, val := i.singleLevelIterator.NextPrefix(succKey); key != nil {

--- a/sstable/reader_iter_two_lvl.go
+++ b/sstable/reader_iter_two_lvl.go
@@ -817,6 +817,11 @@ func (i *twoLevelIterator) NextPrefix(succKey []byte) (*InternalKey, base.LazyVa
 	if key, val := i.singleLevelIterator.NextPrefix(succKey); key != nil {
 		return key, val
 	}
+	// key == nil
+	if i.err != nil {
+		return nil, base.LazyValue{}
+	}
+
 	// Did not find prefix in the existing second-level index block. This is the
 	// slow-path where we seek the iterator.
 	var ikey *InternalKey

--- a/vfs/errorfs/errorfs.go
+++ b/vfs/errorfs/errorfs.go
@@ -9,6 +9,7 @@ import (
 	"io"
 	"os"
 	"strings"
+	"sync/atomic"
 
 	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/errors/oserror"
@@ -188,6 +189,55 @@ func (a anyInjector) MaybeError(op Op) error {
 	}
 	return nil
 }
+
+// Counter wraps an Injector, counting the number of errors injected. It may be
+// used in tests to ensure that when an error is injected, the error is
+// surfaced through the user interface.
+type Counter struct {
+	Injector
+	atomic.Uint64
+}
+
+// String implements fmt.Stringer.
+func (c *Counter) String() string {
+	return c.Injector.String()
+}
+
+// MaybeError implements Injector.
+func (c *Counter) MaybeError(op Op) error {
+	err := c.Injector.MaybeError(op)
+	if err != nil {
+		c.Uint64.Add(1)
+	}
+	return err
+}
+
+// Toggle wraps an Injector. By default, Toggle injects nothing. When toggled on
+// through its On method, it begins injecting errors when the contained injector
+// injects them. It may be returned to its original state through Off.
+type Toggle struct {
+	Injector
+	on atomic.Bool
+}
+
+// String implements fmt.Stringer.
+func (t *Toggle) String() string {
+	return t.Injector.String()
+}
+
+// MaybeError implements Injector.
+func (t *Toggle) MaybeError(op Op) error {
+	if !t.on.Load() {
+		return nil
+	}
+	return t.Injector.MaybeError(op)
+}
+
+// On enables error injection.
+func (t *Toggle) On() { t.on.Store(true) }
+
+// Off disables error injection.
+func (t *Toggle) Off() { t.on.Store(false) }
 
 // FS implements vfs.FS, injecting errors into
 // its operations.

